packages/contracts/src/core/permission/PermissionManager.sol

```diff
@@ -109,6 +182,213 @@ abstract contract PermissionManager is Initializable {
         _initializePermissionManager({_initialOwner: _initialOwner});
     }
 
+    modifier onlyPermissionOwner(
+        address _where,
+        bytes32 _permissionId,
+        uint8 _flags
+    ) {
+        if (_flags == 0) {
+            revert FlagCanNotBeZero();
+        }
+
+        Permission storage permission = permissions[permissionHash(_where, _permissionId)];
+
+        if (_isPermissionFrozen(permission)) {
+            revert PermissionFrozen(_where, _permissionId);
+        }
+
+        // If ROOT is the caller for `grant/revoke/grantWithCondition` functions,
+        // ensure that no owner exists by using counters.
+        if (_isRoot(msg.sender)) {
+            if (
+                (msg.sig == this.grant.selector || msg.sig == this.grantWithCondition.selector) &&
+                permission.grantCounter != 0
+            ) { 
+                revert OwnerExisting(_where, msg.sig, Option.grantOwner);
+            }
+
+            if (msg.sig == this.revoke.selector && permission.revokeCounter != 0) {
+                revert OwnerExisting(_where, msg.sig, Option.revokeOwner);
+            }
+        }
+
+        if (!hasPermission(permission.owners[msg.sender], _flags)) {
```

Comment:
Would be better to get this from storage, since you're accessing it 2nd time in the `revert`.

Reply:
Yeah, the idea was to not hurt the callers who call it correctly.